functionalDB/chapter.md

```diff
@@ -0,0 +1,1094 @@
+    * The output of this chain is a new database element
+    * The _what-if function acts as an adapter from this chain to transaction chain (especially using a database and not an atom and swap!)
+
+As can be seen, this chains differs in the function called by the *_transact* function. This difference is injected into the chain by the function that the user called, either *transact* or *what-if* : 
+
+````clojure
+(defmacro what-if [db & txs]  `(_transact ~db _what-if  ~@txs))
+
+(defmacro transact [db-conn & txs]  `(_transact ~db-conn swap! ~@txs))
+````
+Note that these are not functions, but macros. The reason for using macros here is that arguments to macros do not get evaluated as the call happens, this allows us to offer a cleaner API design where the user provides the operations structured in the same way that any function call is structured in Clojure, and it is up to us to transform the user provided, readable, input to executable transaction / what-if question. 
+
+The main transformation in this process is done by the *_transact* macro. It takes the list of operations and the scenario specific usage function (either *swap!* for a transaction or  *_what-if* for a what-if question) and compose from it a function call to the given scenario specific function. Note that none of these gets evaluated before reaching the body of *transact-on-db.* This is due to the fact the _transact is a macro, and within it, the operations that were provided as function calls (parens wrapped element) are modified into vectors holding the elements of the call (at the *(vec frst-tx#)* call). 
+
+````clojure
+(defmacro  _transact [db op & txs]
+   (when txs
+     (loop [[frst-tx# & rst-tx#] txs  res#  [op db `transact-on-db]  accum-txs# []]
+       (if frst-tx#
+           (recur rst-tx# res#  (conj  accum-txs#  (vec frst-tx#)))
+           (list* (conj res#  accum-txs#))))))
+````
+The role of *_what-if* is to mimic  *swap!* but operate on a database instead of on an Atom.
+
+````clojure
+(defn- _what-if [db f txs]  (f db txs))
+````
+The above described process can be seen in the following examples:
+
+Transaction: 
+
+User call: 
+````clojure
+(transact db-conn  (add-entity e1) (update-entity e2 atr2 val2 :db/add))  
+````
+Changes into: 
+````clojure
+(_transact db-conn swap! (add-entity e1) (update-entity e2 atr2 val2 :db/add))
+````
+Becomes: 
+````clojure
+(swap! db-conn transact-on-db [[add-entity e1][update-entity e2 atr2 val2 :db/add]])
+````
+What-if: 
+
+User call: 
+````clojure
+(what-if my-db (add-entity e3) (remove-entity e4))
+````
+Changes into: 
+````clojure
+(_transact my-db _what-if (add-entity e3) (remove-entity e4))
+````
+Changes into: 
+````clojure
+(_what-if my-db transact-on-db [[add-entity e3] [remove-entity e4]])
+````
+Becomes eventually: 
+````clojure
+(transact-on-db my-db  [[add-entity e3] [remove-entity e4]])
+````
+## Connected data

```

Comment:
This section (everything up until starting the query language), while it has some interesting parts, feels like it takes some of the momentum out of the chapter. After reading about basic CRUD operations and transactional semantics, it feels like the most exciting natural next step would be to go to the query language. 

There's quite a bit of stuff in this chapter, so I wonder how you would feel about possibly omitting this section entirely? We could then mention these two features as exercises to the reader, with the added bonus that there would be a proposed example solution in the code already :)